File: spark-etl/src/main/scala/com/phone/etl/common/ColumnMapping.scala
package com.phone.etl.common

/**
 * 字段中英映射 —— Scala 端的真实源镜像。
 *
 * 与 `scripts/column_mapping.py` 严格保持一致：
 * - Python 端：预处理时把中文列名重命名为英文；
 * - Scala 端：通过本对象把字典刷进 MySQL `ads_column_dict`；
 * - Java/JS 端：从 `ads_column_dict` 反查中文展示。
 *
 * 若新增/修改字段，<b>必须同步更新</b> `scripts/column_mapping.py` 与本文件。
 */
object ColumnMapping {

  /** 普通字段 中文 → 英文 */
  val cnToEn: Map[String, String] = Map(
    "日期"           -> "sale_date",
    "品牌"           -> "brand",
    "型号"           -> "model",
    "处理器性能"     -> "processor",
    "操作系统"       -> "os",
    "促销活动"       -> "promotion",
    "用户评价"       -> "user_rating",
    "客单价"         -> "unit_price",
    "销量"           -> "sales_qty",
    "生产成本"       -> "production_cost",
    "总营销成本"     -> "marketing_cost_total",
    "物流成本"       -> "logistics_cost",
    "总平台佣金"     -> "platform_commission_total",
    "售后服务成本"   -> "after_sales_cost",
    "手机壳销售额"   -> "case_sales",
    "耳机销售额"     -> "earphone_sales",
    "充电器销售额"   -> "charger_sales",
    "数据线销售额"   -> "cable_sales",
    "保护膜销售额"   -> "screen_protector_sales",
    "总配件销售额"   -> "accessory_sales_total",
    "1年延保销售额"  -> "warranty_1y_sales",
    "2年延保销售额"  -> "warranty_2y_sales",
    "意外险销售额"   -> "accident_insurance_sales",
    "碎屏险销售额"   -> "screen_insurance_sales",
    "总延保服务额"   -> "warranty_total_sales",
    "用户所在城市"   -> "user_city",
    "用户年龄"       -> "user_age",
    "用户性别"       -> "user_gender",
    "用户会员等级"   -> "user_member_level"
  )

  /** 拆单位字段：英文 → 中文展示名 */
  val specUnitEnToCnDisplay: Map[String, String] = Map(
    "camera_pixel_wan"     -> "摄像头像素(万)",
    "battery_capacity_mah" -> "电池续航(mAh)",
    "screen_size_inch"     -> "屏幕尺寸(英寸)",
    "storage_gb"           -> "存储容量(GB)",
    "refresh_rate_hz"      -> "屏幕刷新率(Hz)"
  )

  /** Spark 派生 / 聚合列：英文 → 中文展示名 */
  val derivedEnToCn: Map[String, String] = Map(
    "revenue"               -> "营收",
    "total_cost"            -> "总成本",
    "gross_profit"          -> "毛利",
    "gross_margin"          -> "毛利率",
    "marketing_cost_ratio"  -> "营销费用率",
    "accessory_attach"      -> "配件附加率",
    "warranty_attach"       -> "延保附加率",
    "sale_year"             -> "年",
    "sale_month"            -> "月",
    "sale_quarter"          -> "季度",
    "sale_dow"              -> "星期",
    "sale_ym"               -> "年月",
    "age_group"             -> "年龄段",
    "total_revenue"         -> "营收合计",
    "total_qty"             -> "销量合计",
    "avg_unit_price"        -> "平均客单价",
    "total_marketing"       -> "营销费用合计",
    "total_gross_profit"    -> "毛利合计",
    "marketing_ratio"       -> "营销费用率",
    "accessory_sales"       -> "配件销售合计",
    "warranty_sales"        -> "延保销售合计",
    "order_cnt"             -> "订单数",
    "model_cnt"             -> "机型数",
    "avg_user_rating"       -> "平均用户评价",
    "qty_growth_ratio"      -> "销量环比增速",
    "opportunity_score"     -> "机会评分",
    "potential_score"       -> "潜力评分",
    "contribution_ratio"    -> "贡献占比",
    "rolling_margin_30d"    -> "30日滚动毛利率",
    "deviation_ratio"       -> "偏差率",
    "is_anomaly"            -> "是否异常",
    "anomaly_level"         -> "异常级别",
    "metric_code"           -> "指标代号",
    "metric_value"          -> "指标值",
    "mom_ratio"             -> "环比",
    "yoy_ratio"             -> "同比"
  )

  /** 完整 英文 → 中文（供前端反查） */
  val enToCn: Map[String, String] =
    cnToEn.map(_.swap) ++ specUnitEnToCnDisplay ++ derivedEnToCn

  /** 字段所属数仓层 */
  val enToLayer: Map[String, String] = {
    val odsCols = cnToEn.values.toSet ++ specUnitEnToCnDisplay.keySet
    val dwdCols = Set(
      "revenue", "total_cost", "gross_profit", "gross_margin",
      "marketing_cost_ratio", "accessory_attach", "warranty_attach",
      "sale_year", "sale_month", "sale_quarter", "sale_dow", "age_group"
    )
    val dwsCols = Set(
      "total_revenue", "total_qty", "avg_unit_price", "total_marketing",
      "total_gross_profit", "marketing_ratio", "accessory_sales",
      "warranty_sales", "order_cnt", "model_cnt", "avg_user_rating", "sale_ym"
    )
    val adsCols = Set(
      "qty_growth_ratio", "opportunity_score", "potential_score",
      "contribution_ratio", "rolling_margin_30d", "deviation_ratio",
      "is_anomaly", "anomaly_level", "metric_code", "metric_value",
      "mom_ratio", "yoy_ratio"
    )
    enToCn.keys.map { en =>
      val layer =
        if (adsCols.contains(en)) "ads"
        else if (dwsCols.contains(en)) "dws"
        else if (dwdCols.contains(en)) "dwd"
        else if (odsCols.contains(en)) "ods"
        else "unknown"
      en -> layer
    }.toMap
  }

  /** 字段所属业务类目（粗分） */
  val enToCategory: Map[String, String] = {
    def cat(en: String): String = en match {
      case s if Set("sale_date", "sale_year", "sale_month", "sale_quarter",
                    "sale_dow", "sale_ym").contains(s) => "time"
      case s if Set("brand", "model", "processor", "os",
                    "camera_pixel_wan", "battery_capacity_mah",
                    "screen_size_inch", "storage_gb", "refresh_rate_hz").contains(s) => "product"
      case s if Set("user_city", "user_age", "user_gender",
                    "user_member_level", "age_group", "user_rating").contains(s) => "user"
      case s if Set("promotion").contains(s) => "channel"
      case s if s.endsWith("_cost") || s.contains("commission") ||
                s.contains("marketing") => "cost"
      case s if s.endsWith("_sales") || s.startsWith("warranty_") ||
                s.contains("accessory") || s.contains("insurance") => "sales"
      case s if Set("revenue", "total_revenue", "gross_profit", "total_gross_profit",
                    "gross_margin", "unit_price", "sales_qty", "avg_unit_price",
                    "total_qty", "total_cost", "total_marketing", "order_cnt",
                    "model_cnt").contains(s) => "metric"
      case _ => "derived"
    }
    enToCn.keys.map(en => en -> cat(en)).toMap
  }
}

File: app/src/main/java/com/phone/app/common/ColumnMapping.java
package com.phone.app.common;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * 字段中英映射 —— Java 端的真实源镜像。
 *
 * <p>与以下两份保持严格一致（新增/改字段必须同步三处）：
 * <ul>
 *   <li>{@code scripts/column_mapping.py}（预处理脚本）</li>
 *   <li>{@code com.phone.etl.common.ColumnMapping}（Scala/Spark 端）</li>
 * </ul>
 *
 * <p>用途：当 ETL 任务尚未刷新 MySQL `ads_column_dict` 时，前端的字段中文展示
 * 不至于全部空白；同时给本地单测一份兜底。运行时优先以 `ads_column_dict` 为准。
 */
public final class ColumnMapping {

    private ColumnMapping() { /* utility */ }

    /** 英文 → 中文（普通字段 + 派生 + 单位拆分列） */
    public static final Map<String, String> EN_TO_CN;

    /** 英文 → 数仓层（ods/dwd/dws/ads/unknown） */
    public static final Map<String, String> EN_TO_LAYER;

    /** 英文 → 类目（time/product/user/channel/cost/sales/metric/derived） */
    public static final Map<String, String> EN_TO_CATEGORY;

    static {
        Map<String, String> en = new HashMap<>();
        // ods 普通字段
        en.put("sale_date", "日期");
        en.put("brand", "品牌");
        en.put("model", "型号");
        en.put("processor", "处理器性能");
        en.put("os", "操作系统");
        en.put("promotion", "促销活动");
        en.put("user_rating", "用户评价");
        en.put("unit_price", "客单价");
        en.put("sales_qty", "销量");
        en.put("production_cost", "生产成本");
        en.put("marketing_cost_total", "总营销成本");
        en.put("logistics_cost", "物流成本");
        en.put("platform_commission_total", "总平台佣金");
        en.put("after_sales_cost", "售后服务成本");
        en.put("case_sales", "手机壳销售额");
        en.put("earphone_sales", "耳机销售额");
        en.put("charger_sales", "充电器销售额");
        en.put("cable_sales", "数据线销售额");
        en.put("screen_protector_sales", "保护膜销售额");
        en.put("accessory_sales_total", "总配件销售额");
        en.put("warranty_1y_sales", "1年延保销售额");
        en.put("warranty_2y_sales", "2年延保销售额");
        en.put("accident_insurance_sales", "意外险销售额");
        en.put("screen_insurance_sales", "碎屏险销售额");
        en.put("warranty_total_sales", "总延保服务额");
        en.put("user_city", "用户所在城市");
        en.put("user_age", "用户年龄");
        en.put("user_gender", "用户性别");
        en.put("user_member_level", "用户会员等级");

        // 拆单位列
        en.put("camera_pixel_wan", "摄像头像素(万)");
        en.put("battery_capacity_mah", "电池续航(mAh)");
        en.put("screen_size_inch", "屏幕尺寸(英寸)");
        en.put("storage_gb", "存储容量(GB)");
        en.put("refresh_rate_hz", "屏幕刷新率(Hz)");

        // 派生 / 聚合列（覆盖 Spark 端 derivedEnToCn）
        en.put("revenue", "营收");
        en.put("total_cost", "总成本");
        en.put("gross_profit", "毛利");
        en.put("gross_margin", "毛利率");
        en.put("marketing_cost_ratio", "营销费用率");
        en.put("accessory_attach", "配件附加率");
        en.put("warranty_attach", "延保附加率");
        en.put("sale_year", "年");
        en.put("sale_month", "月");
        en.put("sale_quarter", "季度");
        en.put("sale_dow", "星期");
        en.put("sale_ym", "年月");
        en.put("age_group", "年龄段");
        en.put("total_revenue", "营收合计");
        en.put("total_qty", "销量合计");
        en.put("avg_unit_price", "平均客单价");
        en.put("total_marketing", "营销费用合计");
        en.put("total_gross_profit", "毛利合计");
        en.put("marketing_ratio", "营销费用率");
        en.put("accessory_sales", "配件销售合计");
        en.put("warranty_sales", "延保销售合计");
        en.put("order_cnt", "订单数");
        en.put("model_cnt", "机型数");
        en.put("avg_user_rating", "平均用户评价");
        en.put("qty_growth_ratio", "销量环比增速");
        en.put("opportunity_score", "机会评分");
        en.put("potential_score", "潜力评分");
        en.put("contribution_ratio", "贡献占比");
        en.put("rolling_margin_30d", "30日滚动毛利率");
        en.put("deviation_ratio", "偏差率");
        en.put("is_anomaly", "是否异常");
        en.put("anomaly_level", "异常级别");
        en.put("metric_code", "指标代号");
        en.put("metric_value", "指标值");
        en.put("mom_ratio", "环比");
        en.put("yoy_ratio", "同比");
        EN_TO_CN = Collections.unmodifiableMap(en);

        Map<String, String> layer = new HashMap<>();
        EN_TO_CN.keySet().forEach(k -> layer.put(k, inferLayer(k)));
        EN_TO_LAYER = Collections.unmodifiableMap(layer);

        Map<String, String> cat = new HashMap<>();
        EN_TO_CN.keySet().forEach(k -> cat.put(k, inferCategory(k)));
        EN_TO_CATEGORY = Collections.unmodifiableMap(cat);
    }

    private static String inferLayer(String en) {
        switch (en) {
            case "qty_growth_ratio":
            case "opportunity_score":
            case "potential_score":
            case "contribution_ratio":
            case "rolling_margin_30d":
            case "deviation_ratio":
            case "is_anomaly":
            case "anomaly_level":
            case "metric_code":
            case "metric_value":
            case "mom_ratio":
            case "yoy_ratio":
                return "ads";
            case "total_revenue":
            case "total_qty":
            case "avg_unit_price":
            case "total_marketing":
            case "total_gross_profit":
            case "marketing_ratio":
            case "accessory_sales":
            case "warranty_sales":
            case "order_cnt":
            case "model_cnt":
            case "avg_user_rating":
            case "sale_ym":
                return "dws";
            case "revenue":
            case "total_cost":
            case "gross_profit":
            case "gross_margin":
            case "marketing_cost_ratio":
            case "accessory_attach":
            case "warranty_attach":
            case "sale_year":
            case "sale_month":
            case "sale_quarter":
            case "sale_dow":
            case "age_group":
                return "dwd";
            default:
                return "ods";
        }
    }

    private static String inferCategory(String en) {
        switch (en) {
            case "sale_date":
            case "sale_year":
            case "sale_month":
            case "sale_quarter":
            case "sale_dow":
            case "sale_ym":
                return "time";
            case "brand":
            case "model":
            case "processor":
            case "os":
            case "camera_pixel_wan":
            case "battery_capacity_mah":
            case "screen_size_inch":
            case "storage_gb":
            case "refresh_rate_hz":
                return "product";
            case "user_city":
            case "user_age":
            case "user_gender":
            case "user_member_level":
            case "age_group":
            case "user_rating":
                return "user";
            case "promotion":
                return "channel";
            default:
                if (en.endsWith("_cost") || en.contains("commission") || en.contains("marketing")) {
                    return "cost";
                }
                if (en.endsWith("_sales") || en.startsWith("warranty_")
                        || en.contains("accessory") || en.contains("insurance")) {
                    return "sales";
                }
                if (en.equals("revenue") || en.equals("total_revenue") || en.equals("gross_profit")
                        || en.equals("total_gross_profit") || en.equals("gross_margin")
                        || en.equals("unit_price") || en.equals("sales_qty")
                        || en.equals("avg_unit_price") || en.equals("total_qty")
                        || en.equals("total_cost") || en.equals("total_marketing")
                        || en.equals("order_cnt") || en.equals("model_cnt")) {
                    return "metric";
                }
                return "derived";
        }
    }
}
